Fix DateTime(64)-to-Snowflake-and-vice-versa conversion functions

rschu1ze opened the pull request:
Closes #64733
Closes #64713

This PR introduces optional `epoch` parameters for functions `snowflakeToDateTime[64]` and `dateTime[64]ToSnowflake`  before the existing (optional) `time_zone` parameters (*). This makes it technically incompatible for existing usage with `time_zone` parameters. I think this is okay as I don't expect the snowflake functions to be heavily used, in particular not with `time_zone` arguments.

(*) The reason for that is that `time_zone` parameters by convention go last.

This PR contains
- https://github.com/ClickHouse/ClickHouse/pull/64862/commits/5a8be1b9df06e2e28bd0f094a4eb3b1131b2d8c6 - A refactoring to prepare the fixes
- https://github.com/ClickHouse/ClickHouse/pull/64862/commits/17c33d6d96af83cd6274da366fb5592b4009215f - Another refactoring to prepare the fixes even more

- https://github.com/ClickHouse/ClickHouse/pull/64862/commits/7643607cacfb5073edb92ad37b3b059f491c25ab - Fixes #64733
- https://github.com/ClickHouse/ClickHouse/pull/64862/commits/794001445ea377639535991d96e8ecae89fb924a - Fixes #64713

### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Functions `snowflakeToDateTime` and `snowflakeToDateTime64` now accept snowflake IDs of type `UInt64` (previously they required `Int64`). This enables usage with `generateSnowflakeID`, e.g. `SELECT snowflakeToDateTime(generateSnowflakeID())`. Also, `snowflakeToDateTime`, `snowflakeToDateTime64`, `dateTimeToSnowflake`, and `dateTime64ToSnowflake` now default to UNIX epoch (1970-01-01) which makes them consistent with `generateSnowflakeId`. Other epochs, e.g. Twitter's/X's epoch (2010-11-04 aka. 1288834974657 msec since UNIX epoch), can be provided using a new optional `epoch` argument, e.g. `SELECT snowflakeToDateTime(1426860704886947840, 1288834974657)` returns `2021-08-15 10:57:56`.

rschu1ze committed:
Refactorings for consistency, pt. I

robot-ch-test-poll1 commented:
<!-- automatic status comment for PR #64862 from rschu1ze/ClickHouse:melting-some-snowflakes -->
*This is an automated comment for commit c4da339eb3de90105efca8a64a74a2f6534a42b1 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/64862/c4da339eb3de90105efca8a64a74a2f6534a42b1/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>A Sync</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

franz101 commented:
Thanks for the quick rewrite, I'm wondering if the epoch param is part of the standard. For example:
https://github.com/vd2org/snowflake/blob/master/src%2Fsnowflake%2Fsnowflake.py

rschu1ze committed:
Refactorings for consistency, pt. II

rschu1ze committed:
Fix #64733

rschu1ze commented:
@franz101 As far as I see, snowflake IDs are unlike UUIDs not standardized.
- The original Twitter implementation used a 41+10+12 bit layout with epoch 1288834974657 (2010-11-04).
- Discord uses a 41+10+12 bit layout with epoch 1420070400000 (2015-01-01).
- Intragram uses a 41+13+10 bit layout. Parsing the timestamp from such IDs is of course possible, but I did not find the epoch and the format is likely unpopular anyways.
- Mastodon uses 48 bit timestamps with epoch 0 (UNIX epoch, 1970-01-01). The bit layout is too weird to support.
- `generateSnowflakeID` uses a 41+10+12 bit layout with epoch 0 (UNIX epoch).

Yes, it makes sense to add an optional parameter `epoch` to `snowflakeToDateTime`/`snowflakeToDatetime64`. To minimize surprise for users, the default should be epoch 0 of `generateSnowflakeID. This will break compat with existing use cases but I think the function has not been too popular in the first place.

rschu1ze committed:
Fix #64713

antaljanosbenjamin commented on src/Functions/snowflake.cpp:
```diff
@@ -105,32 +101,46 @@ class FunctionSnowflakeToDateTime : public IFunction
+        if (const auto * col_src_const = typeid_cast<const ColumnConst *>(&col_src))
         {
-            Int64 src_val = src_column_const->getValue<Int64>();
+            Int64 src_val = col_src_const->getValue<Int64>();
```
I think this [causes](https://pastila.nl/?001dd8aa/643fda160e29f55954fb727659f5f8a0#aWDe8VXciCutPbZwlrQ29A==) issues with number bigger than int64 max. We should check if it is int64 or uint64.

antaljanosbenjamin commented on src/Functions/snowflake.cpp:
```diff
@@ -96,54 +100,61 @@ class FunctionSnowflakeToDateTime : public IFunction
             {"value", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isUInt64OrInt64), nullptr, "(U)Int64"}
         };
         FunctionArgumentDescriptors optional_args{
+            {"epoch", static_cast<FunctionArgumentDescriptor::TypeValidator>(&isNativeUInt), isColumnConst, "UInt*"},
```
Wouldn't making this epoch the second argument breaks existing queries where the time_zone is the second argument? Wouldn't it be better to have it as third argument?

antaljanosbenjamin commented on src/Functions/snowflake.cpp:
```diff
@@ -49,24 +44,34 @@ class FunctionDateTimeToSnowflake : public IFunction
+        const auto & src_data = typeid_cast<const ColumnUInt32 &>(col_src).getData();
         for (size_t i = 0; i < input_rows_count; ++i)
-            res_data[i] = (Int64(src_data[i]) * 1000 - snowflake_epoch) << time_shift;
+            res_data[i] = (Int64(src_data[i]) * 1000 - epoch) << time_shift;
```
I see that this is a breaking change, however we have to break either the generator function or the converter functions. So I would say it is okay if we list this as a breaking change.

rschu1ze commented:
@antaljanosbenjamin Thanks for checking ... I think I can remove the backward incompatibility by adding two overloads:

```
snowflakeToDateTime(val[, time_zone]) -- as before
snowflakeToDateTime(val[, epoch[, time_zone]]) -- new overload
```

den-crane commented:
>generateSnowflakeID uses a 41+10+12 bit layout with epoch 0 (UNIX epoch).

Logically Clickhouse should use `2024-01-01` and epoch, to extend the usability of generated ids.

rschu1ze commented:
@den-crane: `2024-01-01` seems okay but it is also slightly arbitrary - e.g. why not a more "round" epoch, e.g. `2020-01-01`?.

The default is set to the UNIX epoch 0 because
1. the output of `generateSnowflakeId` is based on the UNIX epoch (yes, this could be changed by adding an epoch parameter to `generateSnowflakeId` ...), and
2. as per the snowflake id "specification" (basically a bunch of random blog posts by a bunch of random guys on the internet), the timestamp component of a snowflake id is the number of milliseconds passed since 1970-01-01 00:00:00.000. In other words, `0` as epoch actually is a natural choice (for sure it is less arbitrary than Twitters/Instragrams/Mastodons/ @den-crane's epochs).

Chances are that we'll have something better than Snowflake ids by 2039. Maybe a 128-bit snowflake. (Maybe we are even are all dead by 2039, given the state of the world. Then none of this will matter anyway.)

---

Regarding backward compatibility: I am torn. I implemented my backward-compatible proposal (https://github.com/ClickHouse/ClickHouse/pull/64862#issuecomment-2152269988) and it is not really pretty, both in terms of implementation and intuitiveness for users. I pushed it separately to have it saved somewhere: https://github.com/ClickHouse/ClickHouse/pull/64916

Another silly proposal would be to keep the existing `snowflakeToDateTime*` and `dateTime*ToSnowflake` functions untouched and add a similar set of functions `snowflakeIDToDateTime*` and `dateTimeToSnowflakeID` which would do it "right", e.g `snowflakeIdToDateTime(val[, epoch[, time_zone]])`. The old functions would be deprecated, hidden behind some setting, and removed at some point.

The thing is that even if we make the call syntax of this PR compatible with existing usage, there will still be an incompatibility due to a different epoch (0 in this PR, Twitter in existing `snowflakeToDateTIme`/`dateTimeToSnowflake`). TO be compatible (at least on demand), means adding another setting which is equally not pretty.

snowflakeToDateTIme`/`dateTimeToSnowflake` is available since two years (https://github.com/ClickHouse/ClickHouse/pull/27704) now. I wrote above that this function is "not heavily used"  ... well I am sure that there is some user who this PR will break.

I am really tempted to implement the silly proposal.

den-crane commented:
I believe that Twitter  (2010-11-04), Discord (2015-01-01) are not accidental, but the year of the implementation.

>2024-01-01 seems okay but it is also slightly arbitrary - e.g. why not a more "round" epoch, e.g. 2020-01-01?.

I don't mind, I just explain the logic, and reasons.

>Snowflake ids by 2039. Maybe a 128-bit snowflake. 

IMHO It completely kills the idea. `snowflakeId` is 64 bit. It's was the only goal of creating snowflakeId. It increase queries performance up to two times in comparison with 128 bits.

And stored size is very nice now:

```
   +-column----+-type------------+-size------+
1. | int64     | UInt64          | 54.52 KiB |
2. | Snowflake | UInt64          | 68.38 KiB |
3. | ulid      | FixedString(26) | 2.51 MiB  |
4. | uuid7     | UUID            | 28.78 MiB |
5. | uuid4     | UUID            | 91.58 MiB |
   +-----------+-----------------+-----------+
```

https://gist.github.com/den-crane/4676ad82e7c6db0db4866ddbd111159b


But I don't insists, just some random minds. I cannot imaging why anyone need to generate any ID in OLAP, it's unthinkable from my POV.

rschu1ze commented:
Continued here: https://github.com/ClickHouse/ClickHouse/pull/64948
